ludwig/hyperopt/execution.py

```diff
@@ -258,15 +280,32 @@ def _cpu_resources_per_trial_non_none(self):
     def _gpu_resources_per_trial_non_none(self):
         return self.gpu_resources_per_trial if self.gpu_resources_per_trial is not None else 0
 
-    def _get_sync_client_and_remote_checkpoint_dir(self, trial_dir: Path) -> Optional[Tuple["CommandBasedClient", str]]:
-        """Get the Ray sync client and path to remote checkpoint directory."""
+    def _get_remote_checkpoint_dir(self, trial_dir: Path) -> Optional[Union[str, Tuple[str, str]]]:
+        """Get the path to remote checkpoint directory."""
         if self.sync_config is None:
             return None
 
-        remote_checkpoint_dir = os.path.join(
-            self.sync_config.upload_dir, *_get_relative_checkpoints_dir_parts(trial_dir)
-        )
-        return get_cloud_sync_client(remote_checkpoint_dir), remote_checkpoint_dir
+        if self.sync_config.upload_dir is not None:
+            # Cloud storage sync config
+            return os.path.join(self.sync_config.upload_dir, *_get_relative_checkpoints_dir_parts(trial_dir))
+        else:
+            # Kubernetes sync config. Returns driver node name and path.
+            # When running on kubernetes, each trial is rsynced to the node running the main process.
+            assert self.kubernetes_namespace is not None
+            node_name = self._get_node_address_by_ip()(self.head_node_ip)
+
+            return (node_name, trial_dir)
+
+    @lru_cache(maxsize=1)
```

Comment:
Can the node address change? I imagine if the ray train job can be made to be fault tolerance to head failure, it could.

Reply:
This is an interesting question. I expect that fault tolerance to head node failure for tune jobs is the only scenario in which head node IP will change.

However, in that case I would expect the `HyperoptExecutor` class to be initialized again, and so would have the correct head node address?

Re: caching here -- the returned value of this function doesn't depend on the head node IP at all, and so even if the head node were to change and we use the cached return val of this function, we'd still have the correct behavior.

Let me know if this makes sense or if there's something I can do to make this more robust.

Reply:
This is an interesting question. I expect that fault tolerance to head node failure for tune jobs is the only scenario in which head node IP will change.

However, in that case I would expect the `HyperoptExecutor` class to be initialized again, and so would have the correct head node address?

Re: caching here -- the returned value of this function doesn't depend on the head node IP at all, and so even if the head node were to change and we use the cached return val of this function, we'd still have the correct behavior.

Let me know if this makes sense or if there's something I can do to make this more robust.

Reply:
I think that sounds reasonable.